[Fix] Support trace-specific color sequences in Plotly Express via templates

antonymilne opened the pull request:
## Summary

Fixes #5416.

You can define color sequences in a template on a per-trace basis via `template.data.<trace_type>`, but these were ignored by Plotly Express. Now they are used. 

## Changes

- Modified `apply_default_cascade()` to check for trace-specific colors in `template.data.<trace_type>` before falling back to `template.layout.colorway` or default qualitative palette
- Added `test_px_templates_trace_specific_colors`

## Example Usage

[See here](https://rawcdn.githack.com/antonymilne/plotly.py/2202662518e9f912c18f26140524eaf54cf12d0e/all_px_template_demo.html) for a demonstration of all (I think?) applicable trace types, i.e. those that support `marker.color`.

```python
import plotly.express as px
import plotly.graph_objects as go

tips = px.data.tips()

# Create a custom template with trace-specific colors
custom_template = go.layout.Template(
    data={
        "histogram": [
            go.Histogram(marker=dict(color="orange")),
            go.Histogram(marker=dict(color="purple")),
        ],
        "scatter": [
            go.Scatter(marker=dict(color="cyan")),
            go.Scatter(marker=dict(color="magenta")),
        ],
    },
    layout=go.Layout(colorway=["yellow", "green"]),
)
template = pio.templates.merge_templates("plotly", custom_template)

# Histogram uses histogram-specific colors
fig1 = px.histogram(tips, x="total_bill", color="sex", template=template)

# Scatter uses scatter-specific colors
fig2 = px.scatter(tips, x="total_bill", y="tip", color="sex", template=template)

# Box plot falls back to layout.colorway
fig3 = px.box(tips, x="day", y="total_bill", color="sex", template=template)
```

<img width="800" alt="changelog_histogram" src="https://github.com/user-attachments/assets/fe988b84-f7b7-4827-8ccf-dc21770ef5f3" />
<img width="800" alt="changelog_scatter" src="https://github.com/user-attachments/assets/42a9c957-3f74-474e-886a-2ae01e98421f" />
<img width="800" alt="changelog_box" src="https://github.com/user-attachments/assets/fed47483-ed00-4009-91ec-72e12aeab6d0" />

## Code PR

- [x] I have read through the [contributing notes](https://github.com/plotly/plotly.py/blob/main/CONTRIBUTING.md) and understand the structure of the package. In particular, if my PR modifies code of `plotly.graph_objects`, my modifications concern the code generator and *not* the generated files.
- [x] I have added tests or modified existing tests.
- [ ] For a new feature, I have added documentation examples (please see the doc checklist as well).
- [x] I have added a CHANGELOG entry if changing anything substantial.
- [ ] For a new feature or a change in behavior, I have updated the relevant docstrings in the code.

emilykl commented on plotly/express/_core.py:
```diff
@@ -1037,9 +1037,30 @@ def apply_default_cascade(args):
+                if not args["color_discrete_sequence"] or not any(
+                    args["color_discrete_sequence"]
+                ):
+                    args["color_discrete_sequence"] = None
```
It would be a bit cleaner to store the colors in a temporary variable and then check if it's valid before assigning it to the args. 

```suggestion
                trace_specific_colors = [
                    trace_data.marker.color
                    for trace_data in trace_data_list
                    if hasattr(trace_data, "marker")
                    and hasattr(trace_data.marker, "color")
                ]
                # If template contains at least one color for this trace type, assign to color_discrete_sequence
                if trace_specific_colors and any(trace_specific_colors):
                    args["color_discrete_sequence"] = trace_specific_colors
```

emilykl commented on plotly/express/_core.py:
```diff
@@ -1003,7 +1003,7 @@ def one_group(x):
 
 
-def apply_default_cascade(args):
+def apply_default_cascade(args, constructor=None):
```
small nit, but I'd prefer not setting a default value for `constructor` here, simply for consistency with the other functions in this file.

`apply_default_cascade(args)` is called in one other place, in `plotly/express/_imshow.py`, so that line should be changed to `apply_default_cascade(args, constructor=None)`

```suggestion
def apply_default_cascade(args, constructor):
```

antonymilne replied:
Great point about consistency with the other functions in the file 👍 Done.

antonymilne commented:
@emilykl thanks very much for your review! I've implemented all your suggested changes exactly as you wrote them, with one small simplification where I changed `if trace_specific_colors and any(trace_specific_colors)` into just `if any(trace_specific_colors)`.

I messed up the commits a bit but you can see all the changes since your last review here: https://github.com/plotly/plotly.py/pull/5437/files/74543f301be86ab121a8069a506ced9aba2a2d53..a3853d2afc113cea2fda3ab98b435e745823abfd.

P.S. if you're in the mood for reviewing, I have another small PR here 😀 https://github.com/plotly/plotly.py/pull/5439

emilykl commented:
@antonymilne One final change needed -- in `plotly/express/_imshow.py`, modify the `apply_default_cascade` call to pass `constructor=None`. I suspect that will fix the failing tests (although I can't see the failure details because CircleCI seems to be having an outage right now). Then this will be ready to merge!

Thanks for your work on this bugfix, and for the additional PR! I will do my best to review tomorrow or early next week.

antonymilne commented:
Oops yes sorry, I made the change to `imshow.py` before but forgot to push it 🤦 

There's a couple of failing tests but I'm pretty sure they're not my fault - I was getting similar messages in #5439. Please could you take a look? 🙏 Maybe they're just flaky and need rerunning? I'm not sure.
